test/fgpu/test_main.py

```diff
@@ -0,0 +1,127 @@
+from katgpucbf.fgpu.output import NarrowbandOutput
+
+
+class TestCommaSplit:
+    """Test :func:`.comma_split`."""
+
+    def test_basic(self) -> None:
+        """Test normal usage, without optional features."""
+        assert comma_split(int)("3,5") == [3, 5]
+        assert comma_split(int)("3") == [3]
+        assert comma_split(int)("") == []
+
+    def test_bad_value(self) -> None:
+        """Test with a value that isn't valid for the element type."""
+        with pytest.raises(ValueError, match="invalid literal for int"):
+            assert comma_split(int)("3,hello")
+
+    def test_fixed_count(self) -> None:
+        """Test with a value for `count`."""
+        splitter = comma_split(int, 2)
+        assert splitter("3,5") == [3, 5]
+        with pytest.raises(ValueError, match="Expected 2 comma-separated fields, received 3"):
+            splitter("3,5,7")
+        with pytest.raises(ValueError, match="Expected 2 comma-separated fields, received 1"):
+            splitter("3")
+
+    def test_allow_single(self) -> None:
+        """Test with `allow_single`."""
+        splitter = comma_split(int, 2, allow_single=True)
+        assert splitter("3,5") == [3, 5]
+        assert splitter("3") == [3, 3]
+        with pytest.raises(ValueError, match="Expected 2 comma-separated fields, received 3"):
+            splitter("3,5,7")
+
+
+class TestParseNarrowband:
+    """Test :func:`.parse_narrowband`."""
+
+    def test_minimal(self) -> None:
+        """Test with the minimum required arguments."""
+        assert parse_narrowband("channels=1024,decimation=8,dst=239.1.2.3+1:7148") == {
+            "channels": 1024,
+            "decimation": 8,
+            "dst": [Endpoint("239.1.2.3", 7148), Endpoint("239.1.2.4", 7148)],
+        }
+
+    def test_maximal(self) -> None:
+        """Test with all valid arguments."""
+        assert parse_narrowband("channels=1024,decimation=8,taps=8,w_cutoff=0.5,dst=239.1.2.3+1:7148") == {
+            "channels": 1024,
+            "decimation": 8,
+            "taps": 8,
+            "w_cutoff": 0.5,
+            "dst": [Endpoint("239.1.2.3", 7148), Endpoint("239.1.2.4", 7148)],
+        }
+
+    @pytest.mark.parametrize(
+        "missing,value",
+        [
+            ("channels", "decimation=8,dst=239.1.2.3+1:7148"),
+            ("decimation", "channels=1024,dst=239.1.2.3+1:7148"),
+            ("dst", "channels=1024,decimation=8"),
+        ],
+    )
+    def test_missing_key(self, missing: str, value: str) -> None:
+        """Test without one of the required keys."""
+        with pytest.raises(ValueError, match=f"--narrowband: {missing} is missing"):
+            parse_narrowband(value)
+
+    def test_duplicate_key(self) -> None:
+        """Test with a key specified twice."""
+        with pytest.raises(ValueError, match="--narrowband: channels specified twice"):
+            parse_narrowband("channels=8,channels=9,decimation=8,dst=239.1.2.3+1:7148")
```

Comment:
Only suggestion I'd make, not that it's particularly important, would be to not have the repeated parameter right next to itself, because that would be more likely what a user would do (IMO).

Reply:
This is a fair point. Please ignore.